### ISSUE TITLE
Different pointers for IDispatch and IUnknown

### ISSUE THREAD

**Issue by cristianbuse** [WaynePhillipsEA/twinbasic#154]
**Describe the bug**
This is not really an issue, I am interested if this is a design decision. In VBA, the ```ObjPtr``` function returns different pointers if given an IDispatch and an IUnknown interface for the same instance while TB shows the exact same pointer.

**To Reproduce**
```VBA
Module Test
    Public Sub Main()
        Dim c As Class1
        Dim iUnk As stdole.IUnknown
        
        Set iUnk = c
        MsgBox ObjPtr(iUnk) - ObjPtr(c)
    End Sub
End Module

Class Class1
End Class
```

**Expected behavior**
In VBA we get a difference between the pointers, in TB we don't.

**Screenshots**
N/A

**Desktop (please complete the following information):**
 - OS: Windows 10 Enterprise V1909
 - twinBASIC compiler v0.9.2720

**Additional context**
None

**Comment by bclothier** [comment]
FWIW from VBA:
```
Public Sub x()
    Dim c As Tester
    Set c = New Tester
    ComparePtrs c
    
    Dim x As Access.Application
    Set x = Application
    ComparePtrs x
End Sub

Private Sub ComparePtrs(Obj As Object)
    Dim Unk As stdole.IUnknown
    Set Unk = Obj
    Debug.Print ObjPtr(Unk) = ObjPtr(Obj), ObjPtr(Unk), ObjPtr(Obj)
End Sub
```

In this case, the `Tester` has different addresses but `Access.Application` has same addresses. This seems to be internal implementation detail, possibly specific to VBA classes.

**Comment by cristianbuse** [comment]
Yes. Also, a Collection also has the same pointer. It must be related to VBA classes.

For example:
```VBA
Module Test
    Public Sub Main()
        Dim c As IClass
        Dim d As Class1
        
        Set c = New Class1
        Set d = c
        
        MsgBox ObjPtr(c) - ObjPtr(d)
        MsgBox ObjPtr(GetDefaultInterface(c)) - ObjPtr(d)
    End Sub

    Private Function GetDefaultInterface(ByVal obj As IUnknown) As Object
        Set GetDefaultInterface = obj
    End Function

End Module

Class Class1
    Implements IClass
End Class

Class IClass
End Class
```

The above code (inTB) shows the exact same thing on both msgbox lines. The same code in VBA would give the second one differently because it would be able to retrieve the default interface. More [here](https://stackoverflow.com/a/63282000/8488913).

Now I wonder how would we get the default interface in TB in this scenario, without casting explicitly to ```Class1```?

**Comment by EduardoVB** [comment]
The ObjPtr function in VB returns different pointers for the same object when obtained from different interfaces.

**Comment by cristianbuse** [comment]
@EduardoVB Yes, but the ```GetDefaultInterface``` function works in VBA regardless of what interface you pass in. It always returns the default one.

**Comment by WaynePhillipsEA** [comment]
Yes, more of an implementation detail, but I agree we should match VB here for ultimate compatibility

**Comment by bclothier** [comment]
I wondered whether this may have to do with the VBA classes implementing the `VBRUN.[_DClass]` so put it to test:

```
Public Sub x()
    Dim x As Class1
    Dim t As Class2
    Dim d As VBRUN.[_DClass]
    Dim u As stdole.IUnknown
    Dim dd As Object
    
    Set x = New Class1
    Set t = x
    Set d = x
    Set u = x
    Set dd = x
    
    Debug.Print "Original:", ObjPtr(x),
    Debug.Print "Impl'd:", ObjPtr(t)
    Debug.Print "_DClass:", ObjPtr(d)
    Debug.Print "IUnknown:", ObjPtr(u)
    Debug.Print "IDispatch:", ObjPtr(dd)
    Debug.Print "Default:", ObjPtr(GetDefaultInterface(t))
End Sub
    
Private Function GetDefaultInterface(ByVal obj As IUnknown) As Object
    Set GetDefaultInterface = obj
End Function
````

In the above code, tested in VBA7, all the pointer addresses were the same, _except_ for the `IUnknown` and implemented interface.  
```
Original:      2688167684864 
Impl'd:        2688167684976 
_DClass:       2688167684864 
IUnknown:      2688167684920 
IDispatch:     2688167684864 
Default:       2688167684864 
Imp'd D:       2688167684864 
```

I verified that in VB6 ActiveX dll, the class also implements the `VBRUN.[_DClass]` and it seems to have the same implementation:

```
Original:      112453744    
Impl'd:        112453804 
_DClass:       112453744 
IUnknown:      112453772 
IDispatch:     112453744 
Default:       112453744 
Imp'd D:       112453744 
```

So I am thinking that any implementation of VB* classes must implicitly implement the `VBRUN.[_DClass]` which apparently acts as the root of all interfaces, and thus enable the `GetDefaultInterface` to work.

**Comment by WaynePhillipsEA** [comment]
This is now fixed in v0.10.5220.  GetDefaultInterface now works, and the internal VBx _DClass interface is also supported.

**Comment by xiaoyaocode163** [comment]
VB6 NO  IUnknown ,SO I USE module.bas to like IUnknown 
every callback like IUnknown ,i also need make a new module file
how to make 10 IUnknown  objext ,and only need one InitializeDemonBlog

Public Function InitializeDemonBlog(LE As DemonBlog) As IUnknown 'IDemonBlog
    If m_pVTable = 0 Then
        With m_VTable
            .VTable(0) = GetAddress(AddressOf QueryInterface)
            .VTable(1) = GetAddress(AddressOf AddRef)
            .VTable(2) = GetAddress(AddressOf Release)
            .VTable(3) = GetAddress(AddressOf ExecuteScriptCompleted)

            m_pVTable = VarPtr(.VTable(0))
        End With
    End If
    
    With LE
        .pVTable = m_pVTable
        CopyMemory InitializeDemonBlog, VarPtr(.pVTable), 4
    End With
End Function
